Implement `java.security.SecureRandom`

armanbilge opened the pull request:
Via `RAND_bytes`, h/t Anton.
https://www.openssl.org/docs/man1.1.1/man3/RAND_bytes.html

Linking to:
- https://github.com/scala-native/scala-native/issues/2600

armanbilge committed:
Implement `java.security.SecureRandom`

lolgab commented on scala-native-crypto/src/java/security/SecureRandom.scala:
```diff
@@ -0,0 +1,22 @@
+
+  override def nextBytes(bytes: Array[Byte]): Unit = Zone { implicit z =>
+    val len = bytes.length
+    val buffer = alloc[Byte](len)
```
We can avoid `Zone` and copying memory by passing the pointer to the byte array directly via
```scala
import scala.scalanative.runtime.ByteArray
bytes.asInstanceOf[ByteArray].at(0)
```

armanbilge replied:
I'd rather not, unless that's an officially supported API (in which case, it should be exposed as a method :)

lolgab replied:
It can't be removed without Scala Native breaking itself. I think it's safe to use. In case it will be removed in future Scala Native versions, we can adapt the code to use the newer alternatives.

keynmol replied:
We can pattern match on subclasses perhaps? https://github.com/scala-native/scala-native/blob/main/nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala#L296

keynmol replied:
To avoid the cast.

armanbilge replied:
I suppose a pattern match would be fine :)

armanbilge replied:
Well, I take that back. Basically my concern is if someone comes up with a new GC scheme, where GC managed memory is not "pinned" and can be moved around.

keynmol replied:
At runtime it's the actual implementation of the array:

<img width="432" alt="image" src="https://user-images.githubusercontent.com/1052965/189343551-dce31b86-1e66-4082-a408-4a4b7682028c.png">

armanbilge replied:
If we feel so strongly about having something like this available, then I argue it should be exposed in Scala Native itself.

keynmol replied:
> Well, I take that back. Basically my concern is if someone comes up with a new GC scheme, where GC managed memory is not "pinned" and can be moved around.

Wouldn't that break Scala Native first? https://github.com/scala-native/scala-native/blob/main/nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala#L167

IMO that's a very big IF.

lolgab replied:
This pattern is used everywhere in the scala-native repository:
![immagine](https://user-images.githubusercontent.com/5793054/189343855-33ab84db-6a83-4cf8-b041-e0554cb7dc53.png)
Also, this library [does this already](https://github.com/lolgab/scala-native-crypto/blob/main/scala-native-crypto/src/java/security/MessageDigest.scala#L139).

armanbilge replied:
Lol, that's comforting 😝

I agree, it's probably fine in practice. But also this is a cryptography library. I'd rather not mess about.

Again, if this is a justified pattern, then let's make a public API and I'll happily make the change :)

keynmol replied:
Yeah I think 

1. If you don't like the cast - pattern match
2. If you are worried about stability of public API changing (and it is public API) - that's IMO something that needs to be revisited when Scala Native breaks it, because that'd need a rewrite of a ton of stuff in scala native itself.

keynmol replied:
> But also this is a cryptography library. I'd rather not mess about

The APIs on ByteArray are public.

armanbilge replied:
> The APIs on ByteArray are public.

Ah, I see your point. `Array[Byte]` can change, but `ByteArray` cannot. That seems good enough for me.

keynmol replied:
When the match fails, you can use either your thing from CE with reusing a small stack buffer, or keep it simple alloc - I'm unaware of scenarious where a byte array will be anything other than `ByteArray` at runtime so it hopefully won't be visible.

That said, a rolling buffer would still be a good "what if" alternative to heap allocation 👍

lolgab replied:
I don't think a library should work forever. If Scala Native breaks we adapt and release a new version. But it's not something I see happening.

lolgab reviewed (changes requested):
Great!
A comment about avoiding allocations.

armanbilge committed:
Avoid allocations

armanbilge committed:
Don't make a CVE ...

armanbilge commented:
Remind me to stop writing security code 😂

lolgab commented:
Noticed that [Scala.js overrides `protected def next(numBits: Int): Int`](https://github.com/scala-js/scala-js-java-securerandom/blob/main/src/main/scala/java/security/SecureRandom.scala#L47).
Do you know why? Should we do the same thing?

Edit: We [probably should](https://github.com/scala-native/scala-native/blob/03e2d5ad13880475524055184f85ea9969657254/javalib/src/main/scala/java/util/Random.scala#L30).

armanbilge commented:
Ouch, very very good catch :(

lolgab committed:
Use cast instead of pattern matching in ByteArray

There is a concern raised by @armanbilge about using an implementation
detail of the Scala Native array implementation (the fact that `Array[Byte]`
are implemented in term of
`scalanative.runtime.ByteArray`
Since relying on this implementation detail is done already in the
library in `MessageDigest`, I'm continuing to rely on it here.
If a future version of Scala Native changes this we will need to update the
library to something different.

lolgab commented:
> Ouch, very very good catch :(

Don't feel sorry about it :) ❤️

lolgab committed:
Port protected def next(Int): Int from Scala.js

lolgab reviewed (approved):
Thank you for the contribution @armanbilge!
